### ISSUE TITLE
python3.8 abi nomenclature change

### ISSUE THREAD

**Issue by ccoulombe** [easybuilders/easybuild-framework#3318]
# Bug/Feature request

With Python 3.8, the abi `m` flag became useless and was removed but this complicate things with `multi_deps` . ( python3.7m -> python3.8 )  https://github.com/python/cpython/commit/7efc526e5cfb929a79c192ac2dcf7eb78d3a4401

For instance, for `configopts` one cannot generically use 
```
-DPYTHON3_INCLUDE_DIR=$EBROOTPYTHON/include/python%(pyshortver)sm"
``` 
It works for python versions under 3.8, but removing the `m` obviously breaks inferior versions.

Using a wildcard (`*`) is not a solution as it is not expanded and the underlying build tool, cmake in this case, needs the full path.

Could a variable such as `%(pyshortver_with_or_without_m)s` be added?

cc : @mboisson

**Comment by mboisson** [comment]
This looks needed to me. Otherwise, every single time we need to refer to the include directory, we need to do it in an easyblock, it can't be done in an easyconfig. 

Thoughts @boegel @Micket ?

**Comment by Micket** [comment]
It might be a bit tricky actually, as one would need to know aoubt the m or lack thereof in https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/framework/easyconfig/templates.py#L250
So, at minimum, it looks like it would be a very intrusive change to me.

**Comment by mboisson** [comment]
Where is the list of templates defined anyway ? I can't find any trace of `pyshortver` being listed in some sort of list of templates.

**Comment by Micket** [comment]
It's right there i linked above.
https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/framework/easyconfig/templates.py#L83
+
https://github.com/easybuilders/easybuild-framework/blob/develop/easybuild/framework/easyconfig/templates.py#L248

**Comment by mboisson** [comment]
Ahh. I was looking for `pyver` and `pyshortver`. So, we would need to have a custom hard-coded case for this ?

**Comment by Micket** [comment]
Yes there is that, but, more problematic is probably how to know whether or not there will be a m suffix in this free standing function (in a way that also doesn't brake testing as a result).

**Comment by mboisson** [comment]
Personally, I have seen : 
2.7 => no m
3.x for x < 8 => m
3.8 => no m

but yeah, that's ugly hard-coding...

**Comment by boegel** [comment]
> For instance, for `configopts` one cannot generically use
> 
> ```
> -DPYTHON3_INCLUDE_DIR=$EBROOTPYTHON/include/python%(pyshortver)sm"
> ```
> 
> It works for python versions under 3.8, but removing the `m` obviously breaks inferior versions.
> 
> Using a wildcard (`*`) is not a solution as it is not expanded and the underlying build tool, cmake in this case, needs the full path.

Why wouldn't `-DPYTHON3_INCLUDE_DIR=$EBROOTPYTHON/include/python%(pyshortver)s*/` work here, it should?

Just like `$EBROOTPYTHON` gets expanded by the shell to which this is passed, so will the `*`?

**Comment by mboisson** [comment]
Good question... probably better asked to a Cmake developer. But it does not get expanded.
  

  
  
  

  
  
>

**Comment by ccoulombe** [comment]
It ends up as :
```
//Python include dir
PYTHON3_INCLUDE_DIR:PATH=/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.6.3/include/python3.6*
```

**Comment by mboisson** [comment]
One suggestion is to add `include/pythonY.X[m]`   to the `CPATH`, in addition to `include`.   Any idea if this might break something ?

**Comment by bartoldeman** [comment]
`python3-config --includes` gives you the correct include paths too.

**Comment by mboisson** [comment]
and `python3-config --abiflags`  gives you the `m` or no-`m`

**Comment by Micket** [comment]
Adding something to CPATH might fix a particular software. Similary, any tolls using the "proper" ways (python3-config) shouldn't even need any help to start with.
I think the real challenge is all the awful build systems out there which expects given paths or such.
Also, things like a custom sanity_check_path in a multi_deps setting would be difficult trying to rely on `python3-config` or the likes.

**Comment by bartoldeman** [comment]
We use PYTHON3_INCLUDE_DIR in the OpenCV easyconfig. Will see what happens there exactly

**Comment by bartoldeman** [comment]
@boegel it seems `run_cmd` via `subprocess.Popen` ends up doing `""`-style shell quoting, so * is not expanded but environment variables are (and also things like `$(python3-config --abiflags)`, so using that instead of `*` avoids the issue.
